src/ptvsd/server/main.py

```diff
@@ -266,62 +266,68 @@ def run_code():
 
 
 def attach_to_pid():
-    def quoted_str(s):
-        if s is None:
-            return s
-        assert not isinstance(s, bytes)
-        unescaped = set(chr(ch) for ch in range(32, 127)) - {'"', "'", "\\"}
-
-        def escape(ch):
-            return ch if ch in unescaped else "\\u%04X" % ord(ch)
-
-        return 'u"' + "".join(map(escape, s)) + '"'
-
-    log.info("Attaching to process with ID {0}", options.target)
+    log.info('Attaching to process with ID {0}', options.target)
 
     pid = options.target
-    host = quoted_str(options.host)
+    host = options.host
     port = options.port
     client = options.client
-    multiprocess = options.multiprocess
-    log_dir = quoted_str(ptvsd.common.options.log_dir)
+    log_dir = common_opts.log_dir
+    if log_dir is None:
+        log_dir = ""
 
-    ptvsd_path = os.path.abspath(os.path.join(ptvsd.server.__file__, "../.."))
-    if isinstance(ptvsd_path, bytes):
-        ptvsd_path = ptvsd_path.decode(sys.getfilesystemencoding())
-    ptvsd_path = quoted_str(ptvsd_path)
+    attach_script_ptvsd_pid_dirname = os.path.join(os.path.dirname(ptvsd.__file__), 'common')
+    assert os.path.exists(attach_script_ptvsd_pid_dirname)
 
-    # pydevd requires injected code to not contain any single quotes.
-    code = """
-import os
-assert os.getpid() == {pid}
 
-import sys
-sys.path.insert(0, {ptvsd_path})
-import ptvsd
-from ptvsd.common import log, options
-del sys.path[0]
+    log_dir = log_dir.replace('\\', '/')
+    setup = {'host': host, 'port': port, 'client': client, 'log_dir': log_dir, 'pid': pid}
 
-options.log_dir = {log_dir}
-log.to_file()
-log.info("Bootstrapping injected debugger.")
+    # We need to be able to import attach_script_ptvsd_pid without importing ptvsd first.
+    if sys.platform == 'win32':
+        setup['pythonpath'] = attach_script_ptvsd_pid_dirname.replace('\\', '/')
 
-debug = ptvsd.attach if {client} else ptvsd.enable_attach
-debug(({host}, {port}), multiprocess={multiprocess})
-""".format(
-        **locals()
-    )
+        python_code = '''import sys;
+sys.path.insert(0, "%(pythonpath)s");
+import attach_script_ptvsd_pid;
+del sys.path[0];
+attach_script_ptvsd_pid.attach(port=%(port)s, host="%(host)s", client=%(client)s, log_dir="%(log_dir)s");
```

Comment:
It feels like this could be simplified if it were to use `execfile()` instead of importing the script. Then you don't need to change `sys.path` at all - just use the full path directly to the script. Values can be passed via `globals=`. Then this whole thing becomes a single-liner:
```py
execfile(..., globals={...})
```
Note that it's called `exec()` in Python 3. This distinction can be handled in `compat`, so that here it's just `compat.execfile()` (can't be `exec()`, because in Python 2 that's a keyword).

Reply:
Another thing - consider how this code handles nested backslashes and quotes in `log_dir`. 

Then there's the issue of string literals, and what happens if e.g. the process is running Python 2, but the code performing the injection is Python 3, and vice versa. Suppose `log_dir` has Unicode characters in it, for example - I don't think it'll survive the trip.

That was why I had that `quote_str()` thing... it takes care of it by making sure that literals in injected code are always explicitly Unicode, `u"..."`, and using `\u` escapes for anything other than ASCII inside, and also for all nested quotes and backslashes.

Reply:
For `compat.execfile` you would still need to load `compat` from `ptvsd.common`. So the `sys.path` dance cannot be avoided. how about this:
```py
import sys
_execfile = execfile if sys.version_info < (3,) else exec
_execfile(<injector-path>, globals={...})
```